### PR TITLE
feat: methods, events to add new rights, scopes

### PR DESCRIPTION
**Pull request by moconnell**

### CODE REVIEW

**Comment by bitbeckers** [comment]
This was a pleasure to review, thanks! I've added some verbose test validations to actually check whether the contract stores the data as expected, but that probably more of a personal pet peeve than necessary.

File: contracts/HypercertMinterV0.sol [hunk 1]
```diff
@@ -91,10 +122,18 @@ contract HypercertMinterV0 is
         require(account != address(0), "Mint: mint to the zero address");
 
         // Parse data to get Claim
-        (Claim memory claim, string memory _uri, bytes32 claimHash) = _bytesToClaimAndURI(data);
+        (Claim memory claim, string memory _uri) = _bytesToClaimAndURI(data);
+
+        for (uint256 i = 0; i < claim.impactScopes.length; i++) {
```
Comment (bitbeckers) [inline contracts/HypercertMinterV0.sol:127]: Note: Check whether passing storage structs to in-contract functions is the most gas efficient pattern. Alternatively, OpenZeppelin also offers Enumerable structs but their data-types might not fit our needs.

P.S. certainly not a showstopper now imho

File: test/setup.ts [hunk 2]
```diff
@@ -1,8 +1,35 @@
...
+};
+
+export type HypercertCollection = {
+  minter: HypercertMinterV0;
+  deployer: AddressedHypercertMinterV0;
+  user: AddressedHypercertMinterV0;
+  anon: AddressedHypercertMinterV0;
+};
+
+const setupTest = deployments.createFixture<
```
Comment (bitbeckers) [inline test/setup.ts:19]: 🙌

File: test/utils.ts [hunk 3]
```diff
@@ -1,79 +1,87 @@
...
+import { ImpactScopes, LoremIpsum, Rights, WorkScopes } from "./wellKnown";
 
-export const getEncodedImpactClaim = async (options?: {
-  rightsIDs?: number[];
+export type Claim = {
+  rights: bigint[] | string[];
+  workTimeframe: number[];
+  impactTimeframe: number[];
+  contributors: string[];
+  workScopes: bigint[] | string[];
```
Comment (bitbeckers) [inline test/utils.ts:13]: Are both types still necessary?

File: test/hypercert_minter/HypercertMinter.minting.ts [hunk 4]
```diff
@@ -140,17 +147,22 @@ export function shouldBehaveLikeHypercertMinterMinting(): void {
...
+    const workScopes = randomScopes(50);
     const options = {
-      rightsIDs: [1],
+      rights: Object.keys(Rights),
       workTimeframe: [1, 2],
       impactTimeframe: [2, 3],
       contributors: [user.address],
-      workScopes: [...Array(50).keys()],
-      impactScopes: [...Array(50).keys()],
+      impactScopes: Object.keys(impactScopes),
```
[the remark this answers is not in the document]
Comment (moconnell) [reply test/hypercert_minter/HypercertMinter.minting.ts:157]: Yep, we're still passing an array of 50 values, I just pulled that bit out into a util function, as we need to actually ensure that those values are now stored in the contract, or else the validation will fail.

```
    const impactScopes = randomScopes(50);
    const workScopes = randomScopes(50);
    const options = {
      rights: Object.keys(Rights),
      workTimeframe: [1, 2],
      impactTimeframe: [2, 3],
      contributors: [user.address],
      impactScopes: Object.keys(impactScopes),
      workScopes: Object.keys(workScopes),
      uri: "ipfs://test",
      version: 0,
    };

    await setupImpactScopes(minter, user.minter, impactScopes);
    await setupWorkScopes(minter, user.minter, workScopes);
```

File: contracts/HypercertMinterV0.sol (hunk 1 above)
Comment (moconnell) [reply contracts/HypercertMinterV0.sol:127]: A quick search turns up comments talking about potential issues from passing around structs with contract size limits, but not much about gas costs. I guess we'd need to do some testing ourselves.

File: test/utils.ts (hunk 3 above)
Comment (moconnell) [reply test/utils.ts]: Removed. Also realised could further tidy using `Partial<T>`.

File: contracts/HypercertMinterV0.sol (hunk 1 above)
Comment (bitbeckers) [reply contracts/HypercertMinterV0.sol:127]: On option is to not use a seperate method and just access the struct directly per for loop

Comment (moconnell) [reply contracts/HypercertMinterV0.sol:127]: ok, inlining the fn does appear to save gas

pre:
hypercerts_pre_inline

post:
hypercerts_post_inline

Comment (bitbeckers) [reply contracts/HypercertMinterV0.sol]: Appreciate the thorough investigation. Indeed it's a little more optimised and I see you refactored methods to inline 👍

File: test/hypercert_minter/HypercertMinter.minting.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bitbeckers) [reply test/hypercert_minter/HypercertMinter.minting.ts:157]: Ah I missed that, apologies. Done :)

**Pull request merged**